libs/core/src/main/java/org/opensearch/core/tasks/TaskId.java

```diff
@@ -106,6 +123,22 @@ public static TaskId readFromStream(StreamInput in) throws IOException {
         return new TaskId(nodeId, in.readLong());
     }
 
+    /**
+     * Read a {@linkplain TaskId} from a byte array. {@linkplain TaskId} has this rather than the usual constructor that takes a
+    * {@linkplain byte[]} so we can return the {@link #EMPTY_TASK_ID} without allocating.
+    */
+    public static TaskId readFromBytes(byte[] in) throws IOException {
+        TaskIdProto.TaskId taskIdProto = TaskIdProto.TaskId.parseFrom(in);
+        if (!taskIdProto.hasNodeId()) {
```

Comment:
Can you help me understand this line of code as I don't see where it's defined in Java, so it may be a generated class.

My thought was that we would read/write a byte array with a VInt length before it, and the array itself gets parsed with protobuf.  If the length is 0 then we skip reading anything and return empty.

If the entire TaskId was optional (see optionalWriteable) this would essentially be the same thing.

But this looks like we are still reading some number of bytes and just determining whether the nodeId index is present; we'd still have 2 bytes for the int even if 0 bytes for the string.

Reply:
The `hasNodeId` method comes from the generated class from protobuf which checks if the optional field nodeId is set or not. For line 132, I kept the same logic as `readFromStream` where if the nodeId is present, it creates a TaskId out of it else it just returns an empty taskId.

Reply:
I think from what I understand, we would be able to achieve the same with keeping both nodeId and id both are optional in the proto message, so similar to the `readFromStream` where it reads the string first as the nodeId and determines if empty taskId should be returned. What do you think?

Reply:
Seems reasonable. Sorry to be so into the details here, but my main concern has been replacing a single-byte solution with anything more complex.   If making both nodeId and id optional results in a 1-byte buffer that works great.

I'm still not clear on how protobuf communicates the total bytes, though.  Looking at #6960 I see the pattern `parseFrom(in.readByteArray())` which implies a length VInt followed by that number of bytes. 

I don't see any equivalent here, or how if everything is optional it provides any signal to the reader that there's nothing to read.  Does protobuf assume reading to end-of-stream?  In that case, reading TaskId would spill over into the TransportMessage.  Or is there an end-of-stream marker?

TLDR: the savings from fewer bytes on the variable length long need to outweigh the extra bytes for field indices and length bytes and any other overhead to make serializing TaskId worth it, and I'm not sure of those numbers here.

Reply:
Totally makes sense! I was thinking about this and we are only writing the bytes when the nodeId is present (https://github.com/opensearch-project/OpenSearch/pull/9737/files#diff-440ec1fd7eb50816eb2d1ec1eb201229b810efcf1939a9953afb0025ffca2f32R154) so I am thinking when we read the bytes, the byte array will have both nodeId and id or nothing at all. Would that not be similar to the current implementation? I might be totally misunderstanding it so would love to know your thoughts.

Also, currently TaskId looks like a separate proto but ultimately it will be part of the OutboundMessageProto with all the bytes and protobuf will deserialize it into various parts.

Reply:
> Totally makes sense! I was thinking about this and we are only writing the bytes when the nodeId is present (https://github.com/opensearch-project/OpenSearch/pull/9737/files#diff-440ec1fd7eb50816eb2d1ec1eb201229b810efcf1939a9953afb0025ffca2f32R154) so I am thinking when we read the bytes, the byte array will have both nodeId and id or nothing at all. Would that not be similar to the current implementation? I might be totally misunderstanding it so would love to know your thoughts.

The issue we're facing here here is that the TaskId is in the middle of the byte stream.   `StreamInput` needs to know where TaskID starts and ends so it can move on to read the bytes for the `TransportRequest`.

Take for example one of the simplest TransportRequests, the initial TCP Handshake Request:
https://github.com/opensearch-project/OpenSearch/blob/5b864c0e023647cf653300d638f60fc0136bf95b/server/src/main/java/org/opensearch/transport/TransportHandshaker.java#L192-L198

Here's what the byte stream looks like (non-protobuf):
```
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 45 53 00 00 00 31 00 00 00 00 00 00 00 08 08 08 |ES...1..........|
|00000010| 20 0b 83 00 00 00 1a 00 00 00 16 69 6e 74 65 72 | ..........inter|
|00000020| 6e 61 6c 3a 74 63 70 2f 68 61 6e 64 73 68 61 6b |nal:tcp/handshak|
|00000030| 65 00 04 a3 8e b7 41                            |e.....A         |
+--------+-------------------------------------------------+----------------+
```
The first 49 bytes (all the way up to the `65` (e) starting the last row) are all part of the `OutboundMessageRequest` header.  So we are receiving the last six bytes that we need to parse into a `TransportMessage`.

The last 5 bytes (`04 a3 8e b7 41`) are the `Writeable` payload for that `HandshakeRequest` class linked above (see `readBytesReference()`, it's a length byte followed by the 4 bytes encoding the version).

That single `00` byte in between the headers and payload is the Task ID.  It's the length of the ID string (0) which shortcuts the 8-byte id.  This is good; when writing if it's "Empty" we just write a 0 byte; when reading when we see a 0 byte we just skip it.

If you "write nothing" how will Protobuf know not to parse it?  There has to be at lease some way to indicate where the TaskID protobuf bytes start and end before we go on to read those last bytes encoding the message.

By the time you get to where `hasNodeId()` works, you've already tried to parse the message into a protobuf, but you don't know where it ends and the "version" in this handshake begins.

I think you need to do something similar to this implementation:
https://github.com/opensearch-project/OpenSearch/blob/5b864c0e023647cf653300d638f60fc0136bf95b/server/src/main/java/org/opensearch/extensions/rest/RegisterRestActionsRequest.java#L38-L47

This reads/writes the protobuf bytes wrapped inside a `ByteArray`.    You can read the byte array and if it is `StreamInput.EMPTY_BYTE_ARRAY` then you can just return the `EMPTY_TASK_ID`.  Otherwise, you take the byte array and call `parseFrom()` on it.

This preserves the 1-byte "empty" behavior we currently have and gives us a slightly shorter (due to variable length task id long) set of bytes (plus the length byte) in the case that we actually have a Task ID.